timelock-impl/src/main/java/com/palantir/atlasdb/timelock/auth/AuthManager.java

```diff
@@ -0,0 +1,125 @@
+import com.palantir.conjure.java.ext.refresh.Refreshable;
+import com.palantir.lock.TimelockNamespace;
+
+public class AuthManager implements AutoCloseable {
+    private final Refreshable<TimelockAuthConfiguration> authConfigurationRefreshable;
+    private final Runnable closingCallback;
+    private volatile AuthServices authServices;
+
+    AuthManager(Refreshable<TimelockAuthConfiguration> authConfigurationRefreshable,
+            Runnable closingCallback,
+            AuthServices authServices) {
+        this.authConfigurationRefreshable = authConfigurationRefreshable;
+        this.closingCallback = closingCallback;
+        this.authServices = authServices;
+    }
+
+    public static AuthManager of(Supplier<TimelockAuthConfiguration> timelockAuthConfigurationSupplier) {
+        PollingRefreshable<TimelockAuthConfiguration> configPollingRefreshable = PollingRefreshable
+                .create(timelockAuthConfigurationSupplier);
+
+        AuthServices authServices = getAuthServices(timelockAuthConfigurationSupplier.get());
+
+        return new AuthManager(
+                configPollingRefreshable.getRefreshable(),
+                configPollingRefreshable::close,
+                authServices);
+    }
+
+    public void checkAuthorized(String clientId, Password password, TimelockNamespace timelockNamespace) {
+        update();
+
+        if (!authServices.useAuth()) {
+            return;
+        }
+
+        Optional<AuthenticatedClient> client = authServices.authenticator().authenticate(clientId, password);
+        if (!client.isPresent() || !authServices.authorizer().isAuthorized(client.get(), timelockNamespace)) {
+            throw new ForbiddenException();
+        }
+    }
+
+    @Override
+    public void close() {
+        closingCallback.run();
+    }
+
+    private void update() {
+        authConfigurationRefreshable.getAndClear().ifPresent(this::updateInternal);
+    }
+
+    private synchronized void updateInternal(TimelockAuthConfiguration authConfiguration) {
+        authServices = getAuthServices(authConfiguration);
+    }
+
+    private static AuthServices getAuthServices(TimelockAuthConfiguration authConfiguration) {
+        Authorizer newAuthorizer = getAuthorizer(authConfiguration);
+        Authenticator newAuthenticator = getAuthenticator(authConfiguration);
+
+        return ImmutableAuthServices.builder()
+                .authenticator(newAuthenticator)
+                .authorizer(newAuthorizer)
+                .useAuth(authConfiguration.useAuth())
+                .build();
+    }
+
+    private static Authorizer getAuthorizer(TimelockAuthConfiguration authConfiguration) {
+        Map<AuthenticatedClient, Privileges> privilegesMap = authConfiguration.privileges().stream()
+                .collect(Collectors.toMap(p -> AuthenticatedClient.create(p.id()), PrivilegesConfiguration::privileges));
+        return SimpleAuthorizer.of(privilegesMap, AuthRequirement.PRIVILEGE_BASED);
```

Comment:
extract out `AuthRequirement.PRIVLEGE_BASED` into a constant explaining that this is what makes timelock 0.x, 1.x etc.